### PR TITLE
Expanded minimum RSA Keylength to 3072

### PR DESCRIPTION
**Pull request by mainboarder**
German Federal Office for Information Security requests in its technical guideline BSI TR-02102-1 RSA Keylength not shorter than 3000bits starting 2024, in the year 2023 3000bits as a recommendation. Gitea should request longer RSA Keys by default in favor of security and drop old clients which do not support longer keys.

https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Publikationen/TechnischeRichtlinien/TR02102/BSI-TR-02102.pdf?__blob=publicationFile&v=9 - Page 19, Table 1.2

### CODE REVIEW

**Comment by wxiaoguang** [comment]
CI fails, the test code also needs to be updated accordingly.

Would you like to fix them or would you like some maintainers to edit this PR to help?

**Comment by mainboarder** [comment]
Searched again and found way more occurences of 2048 bit RSA.
I do need help with: https://github.com/go-gitea/gitea/blob/a4cfeaec9cf6d50a64b98fa01acaa09f2a3146d0/models/asymkey/ssh_key_test.go#L31 and potential other RSA Keys in that file, as I do not know, how the corresponding private key is handeld.

**Review by wxiaoguang — APPROVED** [review]
Vote the approval ahead since it looks good.

While I do not think it is "breaking" because the existing short ssh keys could still be used if I understand correctly.

**Review by delvh — APPROVED** [review]

**Comment by wxiaoguang** [comment]
CI still fails

**Comment by puni9869** [comment]
Unit test need a refactor.

**Comment by puni9869** [comment]
No worries, 

> Honestly dont know what to do. Please help.

Please run `make test` in the console.

File: modules/util/keypair_test.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ import (
 )
 
 func TestKeygen(t *testing.T) {
-	priv, pub, err := GenerateKeyPair(2048)
+	priv, pub, err := GenerateKeyPair(3072)
```
Comment (wxiaoguang) [inline modules/util/keypair_test.go]:
```suggestion
	priv, pub, err := GenerateKeyPair(2048)
```

File: modules/util/keypair_test.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ func TestKeygen(t *testing.T) {
 }
 
 func TestSignUsingKeys(t *testing.T) {
-	priv, pub, err := GenerateKeyPair(2048)
+	priv, pub, err := GenerateKeyPair(3072)
```
Comment (wxiaoguang) [inline modules/util/keypair_test.go]:
```suggestion
	priv, pub, err := GenerateKeyPair(2048)
```

File: models/asymkey/ssh_key_test.go [hunk 3]
```diff
@@ -28,7 +28,7 @@ func Test_SSHParsePublicKey(t *testing.T) {
...
 		{"rsa-1024", false, "rsa", 1024, "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAgQDAu7tvIvX6ZHrRXuZNfkR3XLHSsuCK9Zn3X58lxBcQzuo5xZgB6vRwwm/QtJuF+zZPtY5hsQILBLmF+BZ5WpKZp1jBeSjH2G7lxet9kbcH+kIVj0tPFEoyKI9wvWqIwC4prx/WVk2wLTJjzBAhyNxfEq7C9CeiX9pQEbEqJfkKCQ== nocomment\n"},
-		{"rsa-2048", false, "rsa", 2048, "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDMZXh+1OBUwSH9D45wTaxErQIN9IoC9xl7MKJkqvTvv6O5RR9YW/IK9FbfjXgXsppYGhsCZo1hFOOsXHMnfOORqu/xMDx4yPuyvKpw4LePEcg4TDipaDFuxbWOqc/BUZRZcXu41QAWfDLrInwsltWZHSeG7hjhpacl4FrVv9V1pS6Oc5Q1NxxEzTzuNLS/8diZrTm/YAQQ/+B+mzWI3zEtF4miZjjAljWd1LTBPvU23d29DcBmmFahcZ441XZsTeAwGxG/Q6j8NgNXj9WxMeWwxXV2jeAX/EBSpZrCVlCQ1yJswT6xCp8TuBnTiGWYMBNTbOZvPC4e0WI2/yZW/s5F nocomment"},
+		{"rsa-3072", false, "rsa", 3072, "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDMZXh+1OBUwSH9D45wTaxErQIN9IoC9xl7MKJkqvTvv6O5RR9YW/IK9FbfjXgXsppYGhsCZo1hFOOsXHMnfOORqu/xMDx4yPuyvKpw4LePEcg4TDipaDFuxbWOqc/BUZRZcXu41QAWfDLrInwsltWZHSeG7hjhpacl4FrVv9V1pS6Oc5Q1NxxEzTzuNLS/8diZrTm/YAQQ/+B+mzWI3zEtF4miZjjAljWd1LTBPvU23d29DcBmmFahcZ441XZsTeAwGxG/Q6j8NgNXj9WxMeWwxXV2jeAX/EBSpZrCVlCQ1yJswT6xCp8TuBnTiGWYMBNTbOZvPC4e0WI2/yZW/s5F nocomment"},
```
Comment (wxiaoguang) [inline models/asymkey/ssh_key_test.go]:
```suggestion
		{"rsa-2048", false, "rsa", 2048, "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDMZXh+1OBUwSH9D45wTaxErQIN9IoC9xl7MKJkqvTvv6O5RR9YW/IK9FbfjXgXsppYGhsCZo1hFOOsXHMnfOORqu/xMDx4yPuyvKpw4LePEcg4TDipaDFuxbWOqc/BUZRZcXu41QAWfDLrInwsltWZHSeG7hjhpacl4FrVv9V1pS6Oc5Q1NxxEzTzuNLS/8diZrTm/YAQQ/+B+mzWI3zEtF4miZjjAljWd1LTBPvU23d29DcBmmFahcZ441XZsTeAwGxG/Q6j8NgNXj9WxMeWwxXV2jeAX/EBSpZrCVlCQ1yJswT6xCp8TuBnTiGWYMBNTbOZvPC4e0WI2/yZW/s5F nocomment"},
```

**Review by puni9869 — APPROVED** [review]

**Comment by puni9869** [comment]
🚀

**Comment by wxiaoguang** [comment]
CI passes. And IMO this is not "breaking", all uploaded/generated keys are not affected, so removed the "breaking" label.

Feel free to add the breaking label back if it is really breaking (with detailed breaking description and what users should handle the breaking)

**Comment by mainboarder** [comment]
Thank both of you!

**Comment by silverwind** [comment]
I added a missing `pr/breaking` label to this.

**Pull request merged**